engine/src/main/scala/cromwell/engine/backend/lsf/LsfBackend.scala

```diff
@@ -0,0 +1,261 @@
+  }
+
+  private def updateLsfJobTable(jobDescriptor: BackendCallJobDescriptor, status: String, rc: Option[Int], lsfJobId: Option[Int])
+                               (implicit ec: ExecutionContext): Future[Unit] = {
+    globalDataAccess.updateExecutionInfo(jobDescriptor.workflowDescriptor.id, BackendCallKey(jobDescriptor.call, jobDescriptor.key.index, jobDescriptor.key.attempt), InfoKeys.JobNumber, Option(lsfJobId.toString))
+  }
+
+  /** TODO restart isn't currently implemented for LSF, there is probably work that needs to be done here much like
+    * JES restart, which perhaps could be factored out into a common "remote executor" trait.
+    */
+  override def prepareForRestart(restartableWorkflow: WorkflowDescriptor)(implicit ec: ExecutionContext) = Future.successful(())
+
+  /**
+   * Returns the RC of this job when it finishes.  Sleeps and polls
+   * until the 'rc' file is generated
+   */
+  private def waitUntilComplete(jobDescriptor: BackendCallJobDescriptor): Int = {
+    @tailrec
+    def recursiveWait(): Int = Files.exists(jobDescriptor.returnCode) match {
+      case true => jobDescriptor.returnCode.contentAsString.stripLineEnd.toInt
+      case false =>
+        val logger = jobLogger(jobDescriptor)
+        logger.info(s"'rc' file does not exist yet")
+        Thread.sleep(5000)
+        recursiveWait()
+    }
+    recursiveWait()
+  }
+
+  /**
+   * This returns a zero-parameter function which, when called, will kill the
+   * LSF job with id `lsfJobId`.  It also writes to the 'rc' file the value '143'
+   */
+  private def killLsfJob(jobDescriptor: BackendCallJobDescriptor, lsfJobId: Int) = () => {
+    val bkillStdoutWriter = jobDescriptor.callRootPath.resolve("bkill.stdout").newBufferedWriter
+    val bkillStderrWriter = jobDescriptor.callRootPath.resolve("bkill.stderr").newBufferedWriter
+    val argv = Seq("bkill", lsfJobId.toString)
+    val process = argv.run(ProcessLogger(bkillStdoutWriter writeWithNewline, bkillStderrWriter writeWithNewline))
+    val returnCode: Int = process.exitValue()
+    Vector(bkillStdoutWriter, bkillStderrWriter) foreach { _.flushAndClose() }
+    jobDescriptor.returnCode.clear().appendLine("143")
+    val logger = jobLogger(jobDescriptor)
+    logger.debug(s"bkill $lsfJobId (returnCode=$returnCode)")
+  }
+
+  /**
+   * Writes the script file containing the user's command from the WDL as well
+   * as some extra shell code for monitoring jobs
+   */
+  private def writeScript(jobDescriptor: BackendCallJobDescriptor, instantiatedCommand: String) = {
+    jobDescriptor.script.write(
+      s"""#!/bin/sh
+         |$instantiatedCommand
+         |echo $$? > rc
+         |""".stripMargin)
+  }
+
+  /**
+   * Launches the bsub command, returns a tuple: (rc, Option(lsf_job_id))
+   */
+  private def launchBsub(jobDescriptor: BackendCallJobDescriptor): (Int, Option[Int]) = {
+    val logger = jobLogger(jobDescriptor)
+    val lsfOption = new java.util.HashMap[String,Object]()
```

Comment:
I actually think I preferred the original way of doing this (although I'm sure the assignment could have been generalised into a function).

In particular, I don't like the `java.util.HashMap` here. I personally am not used to mutable collections in the codebase and this one took me by surprise!